vernac/declare.ml

```diff
@@ -107,6 +107,9 @@ type 'a pproof_entry = {
   proof_entry_universes   : UState.named_universes_entry;
   proof_entry_opaque      : bool;
   proof_entry_inline_code : bool;
+  (* True if the proof is declared in an environment which does not
+     contain its side effects and hence should export them. *)
+  proof_leaks_seff        : bool;
```

Comment:
Maybe this could use some rephrasing, I'm not sure I understand it. Is that equivalent to:

"The proof is going to be typechecked in an enviroment where the side effects of the proof are present" ?

Reply:
"...is declared in an environment which does *not* contain its side effects..."

Reply:
they are not there yet, it is up the declaration code to export them

Reply:
Maybe "proof_construction_leaks_effects" (default to true, false when the STM is involved)

Reply:
I see, thanks. What was the reason the STM did rewind the environment?

Reply:
Because, back then, I wanted the code paths of Qed and Defined to be "the same" and Qed needs to rewind since it has to hide the effects (by inlining them). See the message I wrote to Guillaume about the test to refresh your memory on why.

Reply:
I see, so does the linear compiler still need to rewind for `Qed` ?

Reply:
I tried a quick commit to see if I could change that (in the STM) but it did not work out of the box, so I decided to leave this PR as is. It is already large enough imo.

Reply:
Well, if you want to hide the effects yes. But I don't think you need to. This is why I added the extra tests suggested by Guillaume as a separate commit mentioning the STM (see also my answer to his message).

Reply:
Indeed maybe it is just easier to add a table for `Declare.Proof.save` that specs the pre-condition on the caller w.r.t. envs, so at least, we have a spec somewhere (or a desired spec).

Reply:
Once could also have an API to pop from the env a constant under some circumstances, and hide effects that way. My previous answer was considering the APIs we have today.

Reply:
Ok thanks, so if I understand correctly, there is some choice w.r.t. what the callers can do w.r.t. the env, that's fine, but I'd be great if we can document it.

Indeed we don't need to do more stuff in this PR, but I feel that writing down the current invariants could be very helpful.